bolt-modules/boltlib/lib/puppet/functions/resource.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Lookup a resource in the target's data.
+#
+# For more information about resources see [the
+# documentation](https://puppet.com/docs/puppet/latest/lang_resources.html).
+#
+# > **Note:** The `ResourceInstance` data type is under active development and is subject to
+#   change. You can read more about the data type in the [experimental features
+#   documentation](experimental_features.md#resourceinstance-data-type).
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:resource) do
+  # Lookup a resource in the target's data.
+  # @param target The `Target` object to add resources to. See {get_targets}.
+  # @param type The type of the resource
+  # @param title The title of the resource
+  # @return The ResourceInstance if found, or Undef
+  # @example Get the openssl package resource
+  #   $target.apply_prep
+  #   $resources = $target.get_resources(Package).first['resources']
+  #   $target.set_resources($resources)
+  #   $openssl = $target.resource('Package', 'openssl')
+  dispatch :resource do
+    param 'Boltlib::TargetSpec', :target
+    param 'Variant[String[1], Type]', :type
```

Comment:
```suggestion
    param 'Variant[String[1], Type[Resource]]', :type
```

Reply:
Are most custom types a 'Resource' type? I don't think we enforce this when setting resources.

Reply:
We do enforce it for the type's `type` attribute: https://github.com/puppetlabs/bolt/blob/master/bolt-modules/boltlib/lib/puppet/datatypes/resourceinstance.rb#L7 . It looks like that doesn't really apply to `set_resources`, though, since it just expects a hash. Regardless, I believe custom types _should_ be a `Resource` type since they are created using the resource API.

Reply:
We should support `Type[Resource]` but it would probably be better off as a second dispatch of this function rather than a `Variant`.

Reply:
I added that, though for now I don't know if it's safe to convert the string to Type in the `resource_from_string` function, since `set_resource` doesn't do that conversion as well.